lambda/models/scheduling/schedule_management.py

```diff
@@ -614,6 +611,121 @@ def cleanup_scheduled_actions_by_name_pattern(auto_scaling_group: str, model_id:
         logger.error(f"Failed to cleanup scheduled actions by pattern for model {model_id}: {e}")
 
 
+def calculate_next_scheduled_action(schedule_config: SchedulingConfig, timezone_name: str) -> Optional[Dict[str, str]]:
+    """Calculate the next scheduled action (START or STOP) based on the schedule configuration"""
+    try:
+        tz = ZoneInfo(timezone_name)
+        now = datetime.now(tz)
+
+        if schedule_config.scheduleType == ScheduleType.RECURRING:
+            return _calculate_next_recurring_action(schedule_config.recurringSchedule, now, tz)
+        elif schedule_config.scheduleType == ScheduleType.DAILY:
+            return _calculate_next_daily_action(schedule_config.dailySchedule, now, tz)
+
+        return None
+    except Exception as e:
+        logger.error(f"Failed to calculate next scheduled action: {e}")
+        return None
+
+
+def _calculate_next_recurring_action(day_schedule: DaySchedule, now: datetime, tz: ZoneInfo) -> Dict[str, str]:
+    """Calculate next action for recurring schedule"""
+    # Parse schedule times
+    start_hour, start_minute = map(int, day_schedule.startTime.split(":"))
+    stop_hour, stop_minute = map(int, day_schedule.stopTime.split(":"))
+
+    # Create time objects for today
+    today = now.date()
+    start_time = datetime.combine(today, datetime.min.time().replace(hour=start_hour, minute=start_minute)).replace(
+        tzinfo=tz
+    )
+    stop_time = datetime.combine(today, datetime.min.time().replace(hour=stop_hour, minute=stop_minute)).replace(
+        tzinfo=tz
+    )
+
+    # Check which action comes next
+    if now < start_time:
+        # Before start time today - next action is START today
+        return {"action": "START", "scheduledTime": start_time.isoformat()}
+    elif now < stop_time:
+        # Between start and stop today - next action is STOP today
+        return {"action": "STOP", "scheduledTime": stop_time.isoformat()}
+    else:
+        # After stop time today - next action is START tomorrow
+        tomorrow_start = start_time + timedelta(days=1)
+        return {"action": "START", "scheduledTime": tomorrow_start.isoformat()}
+
+
+def _calculate_next_daily_action(
+    daily_schedule: WeeklySchedule, now: datetime, tz: ZoneInfo
+) -> Optional[Dict[str, str]]:
+    """Calculate next action for daily schedule"""
+    current_weekday = now.weekday()
+
+    day_schedules = [
+        daily_schedule.monday,  # 0: Monday
+        daily_schedule.tuesday,  # 1: Tuesday
+        daily_schedule.wednesday,  # 2: Wednesday
+        daily_schedule.thursday,  # 3: Thursday
+        daily_schedule.friday,  # 4: Friday
+        daily_schedule.saturday,  # 5: Saturday
+        daily_schedule.sunday,  # 6: Sunday
+    ]
+
+    # Check today first
+    today_schedule = day_schedules[current_weekday]
+    if today_schedule:
+        next_action = _get_next_action_for_day(today_schedule, now, tz, 0)
```

Comment:
For future design, since `today_schedule` I believe is a `DaySchedule`, it would be a little nicer to have this functionality on that object instead of a function. That concept applies to some other places too.